### DISCUSSION TITLE
Initialising properties with Rust

### DISCUSSION THREAD

**Discussion by iaminapickle** [slint-ui/slint#6995] [Q&A] [answered]
Hi friends!

I want to initialise certain parameters of my app with values I will store in a JSON file and I want to know the best way to do so. I want the app to maintain the same layout (position, size, etc) of certain components as when the user last closed it.

Consider this minimal example:
```
// .slint file
export global FooGlobal {
    in-out property <length> foo_width;
}

component Foo inherits Rectangle {
    in-out property <length> foo_width;
    
    background: gray;
    height: 100%;
    width: foo-width;
}

export component FooWindow inherits Window {
    Foo {}
}
```

I've tried using the `init` callback but the documentation says to not use it for initialisation and also using a setter as shown below but `Foo.width` doesn't seem to change.

```
// .rs file
// Prevent console window in addition to Slint window in Windows release builds when, e.g., starting the app via file manager. Ignored on other platforms.
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use std::error::Error;

slint::include_modules!();

fn main() -> Result<(), Box<dyn Error>> {
    let app = FooWindow::new()?;
    app.global::<FooGlobal>().set_foo_width(100.0);
    
    app.run()?;

    Ok(())
}

```

How would you go about setting `foo-width`? 

Thanks in advance!

**Comment by ogoffart** [comment]
you can also call `app.set_foo_with(100.0)` to initialize the "other" foo-width (the one declared on the component itself)

**Reply by iaminapickle** [reply]
Totally right about having duplicate properties - i had some misunderstandings about how that worked. 

However I do like your approach better but I get this error when i try `app.set_foo_width(100.0)`.

```
error[E0599]: no method named `set_foo_width` found for struct `FooWindow` in the current scope
   --> src/main.rs:11:9
    |
11  |     app.set_foo_width(100.0);
    |         ^^^^^^^^^^^^^ method not found in `FooWindow`
```

**Reply by bjorn** [reply, marked as answer]
Because `FooWindow` doesn't have a `foo_width` property, only `Foo` has. You will need to expose this property on your exported component (and forward it on the Slint side, if needed). Or use the global and bind like `width: FooGlobal.foo_width` instead.
